Adds docs for `dictGet` commands. (#59392)
ClickHouse/ClickHouse: docs/en/sql-reference/functions/ext-dict-functions.md

**Returned value**

If ClickHouse parses the attribute successfully in the [attribute’s data type](../../sql-reference/dictionaries/index.md#dictionary-key-and-fields#ext_dict_structure-attributes), this function returns the value of the dictionary attribute that corresponds to `id_expr`.

If there is no the key, corresponding to `id_expr`, in the dictionary, then this function returns the content of the `<null_value>` element specified for the attribute in the dictionary configuration.

**Implementation details**

ClickHouse throws an exception if it cannot parse the value of the attribute or the value does not match the attribute data type.

**Example**

Query:

```sql
CREATE TABLE source_table
(
    id UInt64,
    value String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO source_table VALUES (1, 'First'), (2, 'Second');

CREATE DICTIONARY id_value_dictionary
(
    id UInt64,
    value String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'source_table'))
LIFETIME(MIN  0 MAX  1000)
LAYOUT(FLAT());

SELECT dictGet('id_value_dictionary', 'value', toUInt64(1)) AS value;
```

**Result**:

```response
'First'
```

## dictGetOrDefault

> Retrieves values from a dictionary. If that value cannot be found, returns the default value set by the user.

Reviewer: I think it's not good to say `If that value cannot be found`, because we search values by keys, and if we cannot find the key, then the default value will be set

Author: I don't understand. Please leave a suggestion.

Reviewer: ```suggestion
Retrieves values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `id_expr` — Key value. [Expression](../../sql-reference/syntax.md#syntax-expressions) returning dictionary key-type value or [Tuple](../../sql-reference/data-types/tuple.md)-type value depending on the dictionary configuration.
- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

If ClickHouse parses the attribute successfully in the [attribute’s data type](../../sql-reference/dictionaries/index.md#dictionary-key-and-fields#ext_dict_structure-attributes), this function returns the value of the dictionary attribute that corresponds to `id_expr`.

If there is no the key, corresponding to `id_expr`, in the dictionary, then returns the value passed as the `default_value_expr` parameter.

**Implementation details**

None.

> 

Reviewer: If none, please, remove this section

Author: I disagree. Pointing out that there aren't any specific implementation details is useful. Otherwise it could look like we just forgot to put the details in (which has happened all throughout our docs).

**Example**

Query:

```sql
CREATE TABLE source_table
(
    id UInt64,
    value String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO source_table VALUES (1, 'First'), (2, 'Second');

CREATE DICTIONARY id_value_dictionary
(
    id UInt64,
    value String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'source_table'))
LIFETIME(MIN  0 MAX  1000)
LAYOUT(FLAT());

SELECT dictGetOrDefault('id_value_dictionary', 'value', toUInt64(1), 'Default return string.') AS value;
```

**Result**:

```response
'Default return string.'
```

## dictGetOrNull

> Retrieves values from a dictionary. If that value cannot be found, returns null. This is different to `dictGet` which throws an exception if the value cannot be found.

Reviewer: ```suggestion
Retrieves values from a dictionary. Returns `NULL` for keys that are not found in the dictionary. This is different to `dictGet` which throws an exception if the key cannot be found.
```

- `id_expr` — Key value. [Expression](../../sql-reference/syntax.md#syntax-expressions) returning dictionary key-type value or [Tuple](../../sql-reference/data-types/tuple.md)-type value depending on the dictionary configuration.

**Returned value**

If ClickHouse parses the attribute successfully in the [attribute’s data type](../../sql-reference/dictionaries/index.md#dictionary-key-and-fields#ext_dict_structure-attributes), this function returns the value of the dictionary attribute that corresponds to `id_expr`.

If there is no the key, corresponding to `id_expr`, in the dictionary, then this fuction returns `null`.

**Implementation details**

None.

**Example**

Query:

```sql
CREATE TABLE source_table
(
    id UInt64,
    value String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO source_table VALUES (1, 'First'), (2, 'Second');

CREATE DICTIONARY id_value_dictionary
(
    id UInt64,
    value String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'source_table'))
LIFETIME(MIN  0 MAX  1000)
LAYOUT(FLAT());

> SELECT dictGetOrNull('id_value_dictionary', 'missing_attribute', toUInt64(1)) AS value;

Reviewer: It will throw an exception `Code: 36. DB::Exception: No such attribute 'missing_attribute': While processing dictGetOrNull('id_value_dictionary', 'missing_attribute', toUInt64(1)) AS value. (BAD_ARGUMENTS)`. Please, test that your examples works.

Functions `OrNull/OrDefault` returns default value/null when the requested key doesn't exist, not when

Author: Suggestions are welcome.

![image](https://github.com/ClickHouse/ClickHouse/assets/9611008/5aa633ab-07b3-4061-ba6e-651daafc3381)

Reviewer: ```suggestion
SELECT dictGetOrNull('id_value_dictionary', 'value', toUInt64(3)) AS value;
```

└──────────────────────────────────────────────────┘
```

## dictGetInt8

Retrieves type Int8 values from a dictionary.

**Syntax**

```sql
dictGetInt8('dictionary_name', 'attribute_name', key)
```

**Arguments**

- `dictionary_name`: the name of the dictionary you want to retrieve data from. [String literal](../syntax#syntax-string-literal)
- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int8` type.

**Implementation details**

> This function will throw an exception if the key does not exist in the dictionary.

Reviewer: This is not true, this function will throw an exception if it cannot parse the value of the attribute or the value does not match the attribute data type, not if the key does not exist.

> This function will throw an exception if the key does not exist in the dictionary.

Reviewer: It will also throw an exception when attribute is not `Int8`

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int8
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Mali', 32);
INSERT INTO countries VALUES (2, 'Czechia', 28);
INSERT INTO countries VALUES (3, 'Brazil', 82);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int8
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetInt8('country_codes', 'code_designation', '3') AS code_designation_output;
```

Result:

```response
28
```

## dictGetInt8OrDefault

> Retrieves type Int8 values from a dictionary. If the Int8 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Int8 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

**Syntax**

```sql
dictGetInt8OrDefault('dictionary_name', 'attribute_name', key, default_value_expr)
```

**Arguments**

- `dictionary_name`: the name of the dictionary you want to retrieve data from. [String literal](../syntax#syntax-string-literal)
- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.
- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int8` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

> This function will throw an exception if the key does not exist in the dictionary.

Reviewer: This is not true, it's `OrDefault` function, it will return default value in this case. But it may throw an excetpion if attribute is not `Int8` type or if default value is not `Int8` or cannot be parsed as `Int8`

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int8
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Mali', 32);
INSERT INTO countries VALUES (2, 'Czechia', 28);
INSERT INTO countries VALUES (3, 'Brazil', 82);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int8
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetInt8OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: This query will throw an exception that it cannot parse `'Default return string.'` as `Int8`

Reviewer: ```suggestion
SELECT dictGetInt8OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

And the same for all other `dictGet<type>OrDefault`

```

Result:

```response
'Default return string.'
```

> ## dictGetInt16

Reviewer: Do we really need this copy paste for every function `dictGet<someType>`? I don't think it's necessary, the are all the same and we can make some general documentation for `dictGet<someType>` and `dictGet<someType>OrDefault`, we can just list supported types and add several examples

Author: Search engine optimization.

Reviewer: What optimizations? I don't see why we need huge copy paste in docs with the same text and examples

Author: If a user searches for `dictGetInt16`, that means they're looking for documentation for the function `dictGetInt16`. Not Int8, not Float32, not anything else. They know what they want, and what they want is `dictGtiInt16`.

In the future, we can automate some of this documentation to populate the changes for us. But right now, the docs team's goal is to document all the holes left by engineers of old. That's what we're doing.

Reviewer: I still think if the user searches for `dictGetInt16` and finds general documentation about `dictGet<type_name>`, it will be enough to understand everything about `dictGetInt16`. For me having 2k lines of documentation for each `dictGet<type_name>` is not a good documentation. But I can be wrong, let me ask what my colleagues think about it.

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int16` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int16
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Sweden', 32767);
INSERT INTO countries VALUES (2, 'Burundi', 28372);
INSERT INTO countries VALUES (3, 'Mozambique', 18372);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int16
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetInt16('country_codes', 'code_designation', '2') AS code_designation_output;
```

Result:

```response
28372
```

## dictGetInt16OrDefault

> Retrieves type Int16 values from a dictionary. If the Int16 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Int16 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int16` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int16
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Sweden', 32767);
INSERT INTO countries VALUES (2, 'Burundi', 28372);
INSERT INTO countries VALUES (3, 'Mozambique', 18372);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int16
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetInt16OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetInt16OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int32` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int32 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'United States', 2147483647);
INSERT INTO countries VALUES (2, 'Maldives', 1938273982);
INSERT INTO countries VALUES (3, 'India', 928372817);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetInt32('country_codes', 'code_designation', '2') AS code_designation_output;
```

Result:

```response
1938273982
```

## dictGetInt32OrDefault

> Retrieves type Int32 values from a dictionary. If the Int32 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Int32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int32` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int32 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'United States', 2147483647);
INSERT INTO countries VALUES (2, 'Maldives', 1938273982);
INSERT INTO countries VALUES (3, 'India', 928372817);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetInt32OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetInt32OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int64` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int64 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Palau', 9223372036854775801);
INSERT INTO countries VALUES (2, 'Jamaica', 384982938291);
INSERT INTO countries VALUES (3, 'Tunisia', 910382938472827168);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int64
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetInt64('country_codes', 'code_designation', '2') AS code_designation_output;
```

Result:

```response
384982938291
```

## dictGetInt64OrDefault

> Retrieves type Int64 values from a dictionary. If the Int64 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Int64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Int64` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation Int64 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Palau', 9223372036854775801);
INSERT INTO countries VALUES (2, 'Jamaica', 384982938291);
INSERT INTO countries VALUES (3, 'Tunisia', 910382938472827168);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int64
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt8` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt8
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Mali', 32);
INSERT INTO countries VALUES (2, 'Czechia', 28);
INSERT INTO countries VALUES (3, 'Brazil', 82);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation UInt8
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetUInt8('country_codes', 'code_designation', '3') AS code_designation_output;
```

Result:

```response
28
```

## dictGetUInt8OrDefault

> Retrieves type UInt8 values from a dictionary. If the UInt8 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type UInt8 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt8` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt8
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Mali', 32);
INSERT INTO countries VALUES (2, 'Czechia', 28);
INSERT INTO countries VALUES (3, 'Brazil', 82);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation UInt8
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetUInt8OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetUInt8OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt16` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt16
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Sweden', 32767);
INSERT INTO countries VALUES (2, 'Burundi', 28372);
INSERT INTO countries VALUES (3, 'Mozambique', 18372);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation UInt16
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetUInt16OrDefault('country_codes', 'code_designation', '999', 'Default value string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetUInt16OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt32` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt32 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'United States', 2147483647);
INSERT INTO countries VALUES (2, 'Maldives', 1938273982);
INSERT INTO countries VALUES (3, 'India', 928372817);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation UInt32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetUInt32('country_codes', 'code_designation', '2') AS code_designation_output;
```

Result:

```response
1938273982
```

## dictGetUInt32OrDefault

> Retrieves type UInt32 values from a dictionary. If the UInt32 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type UInt32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt32` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt32 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'United States', 2147483647);
INSERT INTO countries VALUES (2, 'Maldives', 1938273982);
INSERT INTO countries VALUES (3, 'India', 928372817);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation UInt32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetUInt32OrDefault('country_codes', 'code_designation', '999', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetUInt32OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt64` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt64 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Palau', 9223372036854775801);
INSERT INTO countries VALUES (2, 'Jamaica', 384982938291);
INSERT INTO countries VALUES (3, 'Tunisia', 910382938472827168);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int64
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetInt64('country_codes', 'code_designation', '2') AS code_designation_output;
```

Result:

```response
384982938291
```

## dictGetUInt64OrDefault

> Retrieves type UInt64 values from a dictionary. If the UInt64 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type UInt64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UInt64` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    code_designation UInt64 
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Palau', 9223372036854775801);
INSERT INTO countries VALUES (2, 'Jamaica', 384982938291);
INSERT INTO countries VALUES (3, 'Tunisia', 910382938472827168);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    code_designation Int64
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '2', 'Default return string.') AS code_designation_output;

Reviewer: ```suggestion
SELECT dictGetInt64OrDefault('country_codes', 'code_designation', '999', 42) AS code_designation_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Float32` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    float_designation Float32
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Andora', 1.329382);
INSERT INTO countries VALUES (2, 'Italy', 2.39137);
INSERT INTO countries VALUES (3, 'El Salvador', 4.03928);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    float_designation Float32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetFloat32('country_codes', 'float_designation', '2') AS dict_output;
```

Result:

```response
2.39137
```

## dictGetFloat32OrDefault

> Retrieves type Float32 values from a dictionary. If the Float32 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Float32 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Float32` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    float_designation Float32
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Andora', 1.329382);
INSERT INTO countries VALUES (2, 'Italy', 2.39137);
INSERT INTO countries VALUES (3, 'El Salvador', 4.03928);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    float_designation Float32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetFloat32OrDefault('country_codes', 'float_designation', '999', 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetFloat32OrDefault('country_codes', 'float_designation', '999', 42.42) AS dict_output;
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Float64` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    float_designation Float64
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Madagascar', 1084781.1984689625);
INSERT INTO countries VALUES (2, 'Vanuatu', 109862.3);
INSERT INTO countries VALUES (3, 'Peru', 19474.19);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    float_designation Float32
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetFloat32('country_codes', 'float_designation', '1') AS dict_output;
```

Result:

```response
1084781.2
```

## dictGetFloat64OrDefault

> Retrieves type Float64 values from a dictionary. If the Float64 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Float64 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Float64` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    float_designation Float64
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Madagascar', 1084781.1984689625);
INSERT INTO countries VALUES (2, 'Vanuatu', 109862.3);
INSERT INTO countries VALUES (3, 'Peru', 19474.19);

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    float_designation Float64
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetFloat64OrDefault('country_codes', 'float_designation', '999', 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetFloat64OrDefault('country_codes', 'float_designation', '999', 42.42) AS dict_output;
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Date` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    independence_day Date
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Belarus', '1991-06-03');

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    independence_day Date
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetDate('country_codes', 'independence_day', '1') AS dict_output;
```

Result:

```response
1991-06-03
```

## dictGetDateOrDefault

> Retrieves type Date values from a dictionary. If the Date attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type Date values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.
- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `Date` type. If the attribute cannot be found, the `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    independence_day Date
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Belarus', '1991-06-03');

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    independence_day Date
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetDate('country_codes', 'independence_day', '100', 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetDateOrDefault('country_codes', 'independence_day', '100', '1970-01-01') AS dict_output;
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `DateTime` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt64,
    country String,
    independence_day DateTime
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, 'Belarus', '1970-10-10 00:00:01');

CREATE DICTIONARY country_codes
(
    id UInt64,
    country String,
    independence_day DateTime
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetDateTime('country_codes', 'independence_day', '1') AS dict_output;
```

Result:

```response
1970-10-10 00:00:01
```

## dictGetDateTimeOrDefault

> Retrieves type DateTime values from a dictionary. If the DateTime attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type DateTime values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UUID` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    special_id UUID,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), ' Seychelles');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    special_id UUID,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetUUID('countries_dict', 'special_id', 1) AS dict_output;
```

Result:

```response
0ee3c544-f178-4e93-9112-9fec9a58957b
```

## dictGetUUIDOrDefault

> Retrieves type UUID values from a dictionary. If the UUID attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type UUID values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.
- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `UUID` type. If the attribute cannot be found, the `default_value_expr` is returned.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    special_id UUID,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), ' Seychelles');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    special_id UUID,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetUUIDOrDefault('countries_dict', 'special_id', 999, 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetUUIDOrDefault('countries_dict', 'special_id', 999, '00000000-0000-0000-0000-000000000000') AS dict_output;
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `String` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    unique_id UUID,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), 'Philippines');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    unique_id UUID,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetString('countries_dict', 'country', 1) AS dict_output;
```

Result:

```response
Philippines
```

## dictGetStringOrDefault

> Retrieves type String values from a dictionary. If the String attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type String values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `IPv4` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    unique_id UUID,
    ip_address IPv4,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), '103.202.232.0', 'Spain');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    unique_id UUID,
    ip_address IPv4,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetIPv4('countries_dict', 'ip_address', 1) AS dict_output;
```

Result:

```response
103.202.232.0
```

## dictGetIPv4OrDefault

> Retrieves type IPv4 values from a dictionary. If the IPv4 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type IPv4 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `IPv4` type. If the attribute cannot be found, the value of `default_value_expr` is returned instead.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    unique_id UUID,
    ip_address IPv4,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), '103.202.232.0', 'Spain');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    unique_id UUID,
    ip_address IPv4,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetIPv4OrDefault('countries_dict', 'ip_address', 999, 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetIPv4OrDefault('countries_dict', 'ip_address', 999, '0.0.0.0') AS dict_output;
```

- `attribute_name`: the attribute whose value you want to get. [String literal](../syntax#syntax-string-literal), or [tuple](../data-types/tuple) of column names.
- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `IPv6` type.

**Implementation details**

This function will throw an exception if the key does not exist in the dictionary.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    unique_id UUID,
    ip_address IPv6,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), '0:0:0:0:0:ffff:67ca:e800', 'Spain');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    unique_id UUID,
    ip_address IPv6,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

SELECT dictGetIPv6('countries_dict', 'ip_address', 1) AS dict_output;
```

Result:

```response
::ffff:103.202.232.0
```

## dictGetIPv6OrDefault

> Retrieves type IPv6 values from a dictionary. If the IPv6 attribute cannot be found, returns the specified default value instead.

Reviewer: ```suggestion
Retrieves type IPv6 values from a dictionary. Returns the default value set by the user for keys that are not found in the dictionary.
```

- `key`: the key associated with the attribute. [Expression](../syntax#syntax-expressions) returning dictionary key-type value.
- `default_value_expr` — Values returned if the dictionary does not contain a row with the `id_expr` key. [Expression](../../sql-reference/syntax.md#syntax-expressions) or [Tuple](../../sql-reference/data-types/tuple.md)([Expression](../../sql-reference/syntax.md#syntax-expressions)), returning the value (or values) in the data types configured for the `attr_names` attribute.

**Returned value**

The value of the dictionary attribute that corresponds to the given key. This value is parsed and returned as an `IPv6` type.

**Implementation details**

None.

**Examples**

Query:

```sql
CREATE TABLE countries
(
    id UInt8,
    unique_id UUID,
    ip_address IPv6,
    country String
) ENGINE = MergeTree
ORDER BY id;

INSERT INTO countries VALUES (1, generateUUIDv4(), '0:0:0:0:0:ffff:67ca:e800', 'Spain');

CREATE DICTIONARY countries_dict
(
    id UInt8, 
    unique_id UUID,
    ip_address IPv6,
    country String
)
PRIMARY KEY id
SOURCE(CLICKHOUSE(TABLE 'countries'))
LAYOUT(FLAT())
LIFETIME(MIN 0 MAX 1000);

> SELECT dictGetIPv6OrDefault('countries_dict', 'ip_address', 999, 'Default return string.') AS dict_output;

Reviewer: ```suggestion
SELECT dictGetIPv6OrDefault('countries_dict', 'ip_address', 999, '::ffff:0.0.0.0') AS dict_output;
```
